js/src/components/paid-ads/convertToAssetGroupUpdateBody.test.js

```diff
@@ -0,0 +1,450 @@
+		} );
+	} );
+} );
+
+describe( 'diffAssetOperations', () => {
+	const filledAssetGroup = {
+		[ ASSET_GROUP_KEY.FINAL_URL ]: 'https://example',
+		[ ASSET_GROUP_KEY.DISPLAY_URL_PATH ]: [ 'cap', 'specials' ],
+		assets: {
+			[ ASSET_KEY.BUSINESS_NAME ]: {
+				id: genId(),
+				content: 'General Clothing',
+			},
+			[ ASSET_KEY.MARKETING_IMAGE ]: [
+				{
+					id: genId(),
+					content: 'https://example/image-1.png',
+				},
+			],
+			[ ASSET_KEY.SQUARE_MARKETING_IMAGE ]: [
+				{
+					id: genId(),
+					content: 'https://example/square-image-1.png',
+				},
+			],
+			[ ASSET_KEY.PORTRAIT_MARKETING_IMAGE ]: [
+				{
+					id: genId(),
+					content: 'https://example/portrait-image-1.png',
+				},
+			],
+			[ ASSET_KEY.LOGO ]: [
+				{
+					id: genId(),
+					content: 'https://example/logo-1.png',
+				},
+				{
+					id: genId(),
+					content: 'https://example/logo-2.png',
+				},
+				{
+					id: genId(),
+					content: 'https://example/logo-3.png',
+				},
+			],
+			[ ASSET_KEY.HEADLINE ]: [
+				{
+					id: genId(),
+					content: 'headline 1',
+				},
+				{
+					id: genId(),
+					content: 'headline 2',
+				},
+				{
+					id: genId(),
+					content: 'headline 3',
+				},
+			],
+			[ ASSET_KEY.LONG_HEADLINE ]: [
+				{
+					id: genId(),
+					content: 'long headline 1',
+				},
+			],
+			[ ASSET_KEY.DESCRIPTION ]: [
+				{
+					id: genId(),
+					content: 'description 1',
+				},
+			],
+			[ ASSET_KEY.CALL_TO_ACTION_SELECTION ]: {
+				id: genId(),
+				content: 'sign_up',
+			},
+		},
+	};
+
+	const filledValues = {
+		[ ASSET_FORM_KEY.BUSINESS_NAME ]: 'General Clothing',
+		[ ASSET_FORM_KEY.MARKETING_IMAGE ]: [ 'https://example/image-1.png' ],
+		[ ASSET_FORM_KEY.SQUARE_MARKETING_IMAGE ]: [
+			'https://example/square-image-1.png',
+		],
+		[ ASSET_FORM_KEY.PORTRAIT_MARKETING_IMAGE ]: [
+			'https://example/portrait-image-1.png',
+		],
+		[ ASSET_FORM_KEY.LOGO ]: [
+			'https://example/logo-1.png',
+			'https://example/logo-2.png',
+			'https://example/logo-3.png',
+		],
+		[ ASSET_FORM_KEY.HEADLINE ]: [
+			'headline 1',
+			'headline 2',
+			'headline 3',
+		],
+		[ ASSET_FORM_KEY.LONG_HEADLINE ]: [ 'long headline 1' ],
+		[ ASSET_FORM_KEY.DESCRIPTION ]: [ 'description 1' ],
+		[ ASSET_FORM_KEY.CALL_TO_ACTION_SELECTION ]: 'sign_up',
+	};
+
+	function applyOperations( baseAssetGroup, operations ) {
+		const nextAssets = structuredClone( baseAssetGroup.assets );
+
+		operations.forEach( ( operation ) => {
+			const { field_type: key, id, content } = operation;
+			const isMultiple = Array.isArray( filledAssetGroup.assets[ key ] );
+			let asset = nextAssets[ key ];
+
+			if ( id === null ) {
+				if ( isMultiple ) {
+					if ( ! asset ) {
+						asset = [];
+					}
+					asset.push( { content } );
+				} else {
+					asset = { content };
+				}
+			} else if ( content === null ) {
+				if ( isMultiple ) {
+					asset = asset.filter( ( item ) => item.id !== id );
+				} else {
+					asset = { content: '' };
+				}
+			}
```

Comment:
💅  WDYT of adding a bit of commenting here to explain a bit this nested conditionals?

Reply:
Added in https://github.com/woocommerce/google-listings-and-ads/pull/1874/commits/6ce832a6541fa6e446324fc066200a145d6dc4b3.